types/prosemirror-view/index.d.ts

```diff
@@ -394,7 +394,7 @@ export class EditorView<S extends Schema = any> {
  * them returns true. For some props, the first plugin that yields a
  * value gets precedence.
  */
-export interface EditorProps<S extends Schema = any> {
+export interface EditorProps<T = any, S extends Schema = any> {
```

Comment:
I add to add the `T` type parameter to `EditorProps` so that its functions could specify that the value of their `this` is `Plugin<T, S>`.

Reply:
Suggestion: Make the type parameter more general by using it directly as the type of `this`, instead of restricting the possible types of `this` to types of the shape `Plugin<T, S>`. This would look as follows:

```typescript
export interface EditorProps<ThisT = unknown, S extends Schema = any> {
  handleDOMEvents?: { [name: string]: (this: ThisT, view: EditorView<S>, event: Event) => boolean } | null;
  handleKeyDown?: ((this: ThisT, view: EditorView<S>, event: KeyboardEvent) => boolean) | null;
  // ...
}

export interface PluginSpec<T = any, S extends Schema = any> {
  props?: EditorProps<Plugin<T, S>, S> | null;
  // ...
}

export class Plugin<T = any, S extends Schema = any> {
  // ...
  props: EditorProps<Plugin<T, S>, S>;
  // ...
}
```

Advantages:

* This would remove the dependency on `Plugin` and hence the cycle.
* This change would also make `DirectEditorProps` more accurate: When passing editor props to `new EditorView` we don't know what the value of `this` will be (at least I didn't find anything in the reference manual about this). Therefore in that case the type of `this` should be `unknown`, whereas now it is `Plugin<any, S>`.

Reply:
Ah, you're right @timjb! I hadn't really realized that an EditorView itself can have, for example, its own directly-attached `handleKeyDown()` function that is not a part of any plugin.

Out of curiosity, I tested it to see what the value of `this` would be for that function; it turns out to be the `window` (because it wasn't explicitly bound to anything else). Certainly `unknown` is the right type to specify here, as you suggest.

By the way, I'm also going to bump the TypeScript version number at the top of the files to 3.0 instead of 2.8. I had misread https://github.com/DefinitelyTyped/DefinitelyTyped#older-versions-of-typescript-29-and-earlier, and thought that 2.8 was the currently supported version, but it's actually 3.0, and type `unknown` wasn't introduced until 3.0.